feat: group algebras with sparse representation

thofma opened the pull request:
Due to popular demand, group algebras with sparse representation (aka "don't crash with large groups"). Needs some work/clean up:
```
julia> G = SymmetricGroup(10)
Full symmetric group over 10 elements

julia> QG = group_algebra(QQ, G)
Group algebra
  of full symmetric group over 10 elements
  over rational field

julia> a = QG(rand(G)); b = QG(rand(G));

julia> c = (a + b) * (a - b)
(1,2,4,8,7)(3,9,10,5,6) - (1,4,9)(2,10,3,6,7,5,8) + (1,7,3,8,6,10,4)(2,9,5) - (1,2,6)(3,9)(4,5)
```

@royess this might interest you. It is similar to your approach.

Small todo:
- [x] Add a `FG(::Dict)` constructor.
- [x] Add tests.
- [x] Add documentation.

lgoettgens commented:
Does this change allow me to work with group algebras of infinite groups as well?

thofma commented:
Yes:
```
julia> F = free_group(2)
Free group of rank 2

julia> x, y = gens(F);

julia> QF = group_algebra(QQ, F; sparse = true)
Group algebra
  of free group of rank 2
  over rational field

julia> (QF(x) + QF(y))^2
f1^2 + f1*f2 + f2*f1 + f2^2
```

Fe-r-oz commented:
This is wonderful! I will shift to `sparse = true` as soon as this is merged! Thank you.

fingolfin commented on src/AlgAss/Types.jl:
```diff
@@ -130,6 +130,13 @@ end
 @attributes mutable struct GroupAlgebra{T, S, R} <: AbstractAssociativeAlgebra{T}
   base_ring::Ring
   group::S
+  # We represent elements using a coefficient vector,
```
```suggestion
  # If `sparse` is false we represent elements using a coefficient vector,
```

thofma replied:
I mean vector in the mathematical sense. They comment there is valid both for sparse and non-sparse representations.

fingolfin commented on src/AlgAss/Types.jl:
```diff
@@ -226,8 +272,22 @@ mutable struct GroupAlgebraElem{T, S} <: AbstractAssociativeAlgebraElem{T}
 end
 
+__elem_index(A, g) = get!(A.group_to_base, g) do
+        A.ind += 1
+        resize!(A.base_to_group, max(A.ind, length(A.base_to_group)))
+        A.base_to_group[A.ind] = g
+        return A.ind
```
So when would `length(A.base_to_group)` ever exceed `A.ind`? Right now I am wondering why you even need `A.ind` and can't just do this:
```suggestion
        push!(A.base_to_group, g)
        return length(A.base_to_group)
```

thofma replied:
Yes, I think I wanted to do work with A.base_to_group thing differently, but it does not make any sense now. Will do as you suggest.

fingolfin commented on src/AlgAss/Types.jl:
```diff
@@ -146,77 +153,116 @@ end
-      v[1] = one(K)
 
-      A.one = v
+      if !A.sparse
```
Merge this with preceding `end` into an `else`?

fingolfin reviewed (commented):
Great! Some quick comments and questions (all nitpicks, not meant to block anything)

thofma commented on docs/src/manual/algebras/groupalgebras.md:
```diff
@@ -0,0 +1,69 @@
+
+julia> QG = group_algebra(QQ, G);
+
+julia> QG(Dict(a => 2, zero(G) => 1)) == 2 * QG(a) + 1 * QG(zero(G))
```
added

thofma commented on src/AlgAss/AlgGrp.jl:
```diff
@@ -59,17 +59,25 @@ end
 
 @doc raw"""
-    group_algebra(K::Ring, G; op = *) -> GroupAlgebra
+    group_algebra(K::Ring, G::Group; sparse::Bool = false,
```
I got rid of it by adding another layer of indirection.

thofma commented:
Thanks for the comments. I think I addressed them all. I also removed the `sparse` option from the user-facing version. It is now chosen automatically (as in Magma).
